cmd/soroban-rpc/lib/preflight/src/ledger_storage.rs

```diff
@@ -227,28 +245,27 @@ impl LedgerStorage {
 
 impl SnapshotSource for LedgerStorage {
     fn get(&self, key: &Rc<LedgerKey>) -> Result<(Rc<LedgerEntry>, Option<u32>), HostError> {
-        let mut entry_and_expiration =
-            <LedgerStorage>::get(self, key, self.restore_tracker.is_some())?;
         if let Some(ref tracker) = self.restore_tracker {
+            let mut entry_and_expiration = self
+                .get_including_persistent_expired(key)
+                .map_err(HostError::from)?;
             // If the entry expired, we modify the expiration to make it seem like it was restored
             entry_and_expiration.1 =
                 tracker.track_and_restore(self.current_ledger_sequence, key, &entry_and_expiration);
+            return Ok((entry_and_expiration.0.into(), entry_and_expiration.1));
         }
-        Ok((entry_and_expiration.0.into(), entry_and_expiration.1))
+        let entry_and_expiration =
+            <LedgerStorage>::get(self, key, false).map_err(HostError::from)?;
+        return Ok((entry_and_expiration.0.into(), entry_and_expiration.1));
     }
 
     fn has(&self, key: &Rc<LedgerKey>) -> Result<bool, HostError> {
-        let entry_and_expiration =
-            match <LedgerStorage>::get(self, key, self.restore_tracker.is_some()) {
-                Err(e) => match e {
-                    Error::NotFound => return Ok(false),
-                    _ => return Err(HostError::from(e)),
-                },
-                Ok(le) => le,
-            };
-        if let Some(ref tracker) = self.restore_tracker {
-            _ = tracker.track_and_restore(self.current_ledger_sequence, key, &entry_and_expiration);
+        let result = <dyn SnapshotSource>::get(self, key);
+        if let Err(ref host_error) = result {
+            if host_error.error == HostError::from(Error::NotFound).error {
+                return Ok(false);
+            }
         }
```

Comment:
I am not sure this is the best way (or even a correct way) to compare Host errors

@dmkozh / @leighmcculloch  can you take a look?

Reply:
Typically we verify host errors wit hhost_error.error.is_code/is_type, but for that you'll need to know the mapping from Error::NotFound.

Reply:
So, is the code correct?

Reply:
I found a way to improve it. Thanks